Fix part of #8423 :Fixed the test checking period at the end of docstring.

tanishq67 opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes or fixes part of #8423.
2. This PR does the following: It fixes the lint test written to check period at the end of docstring,

## Essential Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## Proof that changes are correct

<!--
Add videos/screenshots of the user-facing interface to demonstrate that the changes made in this PR work correctly.
-->

## PR Pointers

- Make sure to follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#instructions-for-making-a-code-change).
- Oppiabot will notify you when you don't add a PR_CHANGELOG label. If you are unable to do so, please @-mention a code owner (who will be in the Reviewers list), or ask on [Gitter](https://gitter.im/oppia/oppia-chat).
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Never force push. If you do, your PR will be closed.
- Oppiabot can assign anyone for review/help if you leave a comment like the following: "{{Question/comment}} @{{reviewer_username}} PTAL"
- Some of the e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-your-build-fails).

tanishq67 committed:
Fixed the test checking period at the end of docstring.

tanishq67 commented:
Hello @sajalasati, the solution I proposed was not so simple to implement because-
The function `visit_classdef` was supposed to be obtaining `classdef` node but the tests were written in such a way that it was passing `module` node to `visit_classdef()` and working with that, however when the actual code will run it will obviously pass `classdef node`.
I initially thought that it was obtaining the classdef node as writen in the docstring and implemented the simple solution which I proposed before, then I realised it wasnt working and when I dig deeper into it I figured out that it was mentioned `classdef` however it is the `module node` which was being to it. So, inorder to work with my solution I have made `visit_classdef` function which is taking classdef nodes and named the previous function `visit_module` ( and hence it's corresponding tests are calling visit_module function). I have also written the tests and now **period at the end** is working fine and **rest of the code is untouched.**

tanishq67 committed:
fixed all the lint errors in the current code base.

DubeySandeep commented:
@sajalasati Can you please take the first pass and assign other codeowners after approving the PR? :)

sajalasati commented on scripts/linters/pylint_extensions.py:
```diff
@@ -589,12 +589,12 @@ class DocstringParameterChecker(checkers.BaseChecker):
     DOCSTRING_SECTION_YIELDS = 'yields'
     DOCSTRING_SECTION_RAISES = 'raises'
 
-    def visit_classdef(self, node):
```
Do you want to say the check was written for module node rather than Class node, as you didn't change rest of the code? I don't think so.

sajalasati commented on scripts/linters/pylint_extensions_test.py:
```diff
@@ -335,7 +335,7 @@ class ClassName(dummy_class):
         node_no_newline_below_class_docstring.file = filename
         node_no_newline_below_class_docstring.path = filename
 
-        self.checker_test_object.checker.visit_classdef(
```
If the tests were wrongly written you should fix them and test the code to see the source of bug, not rename the check based on tests. Also you said the actual code would be passing Class node not module node, then why doesn't the current checker catches those?

sajalasati reviewed (changes requested):
@tanishq67 Thanks for your PR! Left comments. PTAL!
I didn't exactly get why you renamed the check function. I see this as a hack and not a proper solution, and there are some other issues too. Please try to understand why the current check isn't catching the errors first.

sajalasati commented:
@tanishq67 Currently you are fixing this bug for Class docstrings, but can you also inspect the same for Functions, and if it exists there then fix that too? You can write about your inspection in PR description. Thanks!

sajalasati commented:
Hey @tanishq67! Any updates on this PR? I did some analysis related to the task this PR is solving. You can take a look at #12327 to get more idea. Thanks!

tanishq67 commented:
Hi, @sajalasati Thank you very much for you remarks. 

> @tanishq67 Currently you are fixing this bug for Class docstrings, but can you also inspect the same for Functions, and if it exists there then fix that too? You can write about your inspection in PR description. Thanks!

The test is working absolutely fine with the Functions.
I initially got confused after reading the tests and hence changed the name of visit_classdef() to visit_module(). I went through it again as per your suggestion. Now I think if I will transfer the code currently in `visit_module()` to the `visit_classdef()`, it should work absolutely fine. 
However, The problem is that, we have a module node in the test and I'm not able to extract `classdef` nodes from it to pass to the `visit_classdef` function. Can you help me with this?

sajalasati commented:
> Hi, @sajalasati Thank you very much for you remarks.
> 
> > @tanishq67 Currently you are fixing this bug for Class docstrings, but can you also inspect the same for Functions, and if it exists there then fix that too? You can write about your inspection in PR description. Thanks!
> 
> The test is working absolutely fine with the Functions.
> I initially got confused after reading the tests and hence changed the name of visit_classdef() to visit_module(). I went through it again as per your suggestion. Now I think if I will transfer the code currently in `visit_module()` to the `visit_classdef()`, it should work absolutely fine.
> However, The problem is that, we have a module node in the test and I'm not able to extract `classdef` nodes from it to pass to the `visit_classdef` function. Can you help me with this?

@tanishq67 Sorry I still do not get what your approach is. Please explain why you need `visit_module()` when you already have `visit_classdef` in the existing [code](https://github.com/oppia/oppia/blob/450c958fb422223667962eb9d50d27022be2cd85/scripts/linters/pylint_extensions.py#L592)? Your task is to look at the current code we have, identify why we are missing this lint check, and then implement/fix it.

Also can you edit the PR title, as this is not about fixing a backend test (which I think you are trying to do currently), it's about fixing the lint check itself. Also PTAL at my previous [comment](https://github.com/oppia/oppia/pull/12279#issuecomment-806399038) it would help you understand this issue and arrive at the solution. Thanks!

tanishq67 committed:
Merge remote-tracking branch 'upstream/develop' into period

sajalasati commented:
@tanishq67 Would you be able to complete this PR? Else I am working on a PR to solve #12327 which will cover this PR also as a subtask. Kindly let me know. Thanks!

tanishq67 commented:
@sajalasati I will update the PR by tonight, also I assure you this time I will clear all the confusions.

tanishq67 commented:
@sajalasati I was working on it, and its nearly done. I will push it tommorow, thankyou for the patience.

tanishq67 committed:
Fixed all the tests

tanishq67 committed:
Merge remote-tracking branch 'upstream/develop' into period

tanishq67 committed:
Fixed all the tests

tanishq67 commented:
I will try to explain you the problem and how I approached it and why now it is solved also please visit the refrences I linked they will help you understand my point in much more detail.

So, the issue came to light due to [this](https://github.com/oppia/oppia/pull/11910#discussion_r574517696) error not being caught, now when I searched and analysed it.
I came to find out that appranetly the **lint test** is only working for those classes which have a constructor, ( [ref](https://github.com/oppia/oppia/issues/8423#issuecomment-798878041) ) now, inorder to counter that I simply solved it by directlly passing the `classdef node` rather than `check_docstring_style()`. 

The code written previously in `classdef function` was perfect, but the tests written for them were not so correct reason being that if you will look carefully ( [ref](https://github.com/oppia/oppia/blob/7caa28c2c9a5b4229edc088f1ad69a9afb4da73a/scripts/linters/pylint_extensions_test.py#L321) ), you will realise that it is passing `module node` rather than a `classdef node` which is absolutley incorrect.


How I wrote the code for tests which fixed them -
If you will take a look at `classdef function` it is having ( [ref](https://github.com/oppia/oppia/blob/13ff3a60f6d940791466b108fe149b7a131140e2/scripts/linters/pylint_extensions.py#L606) ) `node.root().file`, if I will pass a classdef node with no module node then the visit_classdef() will run into infinite loop. 
( take a look at the while true [ref](https://github.com/oppia/oppia/blob/13ff3a60f6d940791466b108fe149b7a131140e2/scripts/linters/pylint_extensions.py#L605)),  However, it will work in **actual case**(when the Lint tests will run in real time) since actual does have the `module node` for every `classdef node`. So, all that is needed was to somehow replicate the actual scenerio in the test.
For that I wrote the code in such a way that now I have my `classdef node` **having a module and a file**.
This made the tests run perfectly :).

Concluding everything  -
Previous code was perfect for `visit_classdef()`, but no so great for this function `check_docstring_style()` and tests involving `visit_classdef() function`.
I implemented a simple alternative for `check_docstring_style()`.
Tests involving visit_classdef() were passing `module node` rather than `classdef node` prevoiusly, I cured them by passing `classdef node` in such a way that they have a `module node` and hence we will not run into any error as `nood.root().file` do exist now.
Now, the **lint test period at the end** and **other lint test** checked in `check_docstring_structure()` are working fine for the `classes` **regardless if they have a constructor or not**. ( Due to this reason in my PR I have also fixed the lint errors present in the existing codebase caught by these lint tests. This proves that they are working now. ) 

I think this time I was clearer than ever, if in case you have any suggestions or you think that I have done something incorrectly please let me know I will try to fix it.
Again I cannot thank you enough in words for the patience and support you gave me. Thank you very much.

tanishq67 commented:
@sajalasati PTAL, and do read my [comment](https://github.com/oppia/oppia/pull/12279#issuecomment-812754913). Thanks!

sajalasati commented on scripts/linters/pylint_extensions_test.py:
```diff
@@ -319,21 +319,23 @@ def setUp(self):
+            test_code)
+
+        module_no_newline_below_class_docstring = (
+            node_no_newline_below_class_docstring.root())
```
Doesn't this mean that you are again converting it into a module node? I think that's what using `.root()` does here. If that's true then how is it different from existing situation? (Just FYI, you can access file and filename directly also like `module_no_newline_below_class_docstring.root().file` without converting the node itself to its root)

I think we can also find a way to reach the start of docstring without using a for loop (check [here](http://pylint.pycqa.org/projects/astroid/en/latest/api/astroid.nodes.html#astroid.nodes.FunctionDef)), and then we won't need to write into a temp file and specify file, filename etc for the node, making the tests much more simpler where you only need to create a classDef node.

Also can you provide some proof of how you verified that tests weren't working properly earlier, and after you made changes they did? (Maybe print something on the terminal to demonstrate that, and show us a screenshot of it)

tanishq67 replied:
No, previously we were passing `module node` to the `visit_classdef()` however, this time the `module node` is being made in order to replicate the actual scenario only and `classdef node` is only being passed to the `visit_classdef()`.

 I already checked the documentation but and I think the current method might be the only method to get the raw source code.

sajalasati replied:
> No, previously we were passing module node to the visit_classdef() however, this time the module node is being made in order to replicate the actual scenario only and classdef node is only being passed to the visit_classdef().

Ohh sorry I got confused with variable names a bit, now I see that both are separate. Thanks for clarifying!

> I already checked the documentation but and I think the current method might be the only method to get the raw source code.

Actually there is a way. Take a look at `pylint_extensions.py` file in #11450 - you can see the author has used a trick to understand where the docstring is ending using by splitting node.doc - see [here](https://github.com/oppia/oppia/pull/11450/files#diff-900df16033a01355552cd3477b8fa114f0fbb1506fb740207a7ec7a803009aacR1918).

tanishq67 replied:
>Actually there is a way. Take a look at pylint_extensions.py file in #11450 - you can see the author has used a trick to understand where the docstring is ending using by splitting node.doc - see here.

@sajalasati I did thought about doing something like that but unfortunately it will not work, the author of that PR has assumed that the function's arguments will not extent more than a line.
Just check out this ss, I have replicated what he is doing...
![image](https://user-images.githubusercontent.com/52757132/113655996-930a3280-96b8-11eb-8945-db619b455bde.png)
you got my point?

sajalasati replied:
@tanishq67 The PR author did handle that case, please look at the full code (I think line 1919-1921 covers the case).

tanishq67 replied:
>Also can you provide some proof of how you verified that tests weren't working properly earlier, and after you made changes they did? (Maybe print something on the terminal to demonstrate that, and show us a screenshot of it)

I added this statement to print the node.doc 
![Screenshot from 2021-04-15 10-03-03](https://user-images.githubusercontent.com/52757132/114815243-75814b00-9dd3-11eb-8c76-3bccbd01ec4d.png)
Checkout the result
![Screenshot from 2021-04-15 10-14-48](https://user-images.githubusercontent.com/52757132/114815302-8f229280-9dd3-11eb-8ca1-977740d2d45a.png)
It is printing the doc of module node - 
![Screenshot from 2021-04-15 10-04-21](https://user-images.githubusercontent.com/52757132/114815372-b11c1500-9dd3-11eb-886d-1e27d1cd964e.png)

Is this convincing ?

tanishq67 replied:
> @tanishq67 The PR author did handle that case, please look at the full code (I think line 1919-1921 covers the case).

@sajalasati I think we can do that. I'll try and if it works I'll update all the ten tests, else I'll let you know. Thanks!

sajalasati replied:
> Is this convincing ?

@tanishq67 These screenshots only shows that execution reached the print statement line in your code, but how exactly do you demonstrate that earlier test wasn't working properly and the modified one does?

And how it it possible that those tests were passing earlier if they were passing the wrong node only? Go through the old code again and make sure you properly understand how it works.

Regarding what you are expected to do - you should probably first decide the logic that should be used inside `visit_classdef` function so that both the old checks work as well as the new check can be added, and then modify old tests/write new tests according to that logic. Is it clear now?

tanishq67 replied:
>@tanishq67 These screenshots only shows that execution reached the print statement line in your code, but how exactly do you demonstrate that earlier test wasn't working properly and the modified one does?

@sajalasati I was trying to show that the tests are passing module node since, otherwise it should have printed this -
(highlighted text)
![image](https://user-images.githubusercontent.com/52757132/114824352-635ad900-9de2-11eb-9db4-c5bd1e648e69.png)


in the node.doc , dont you think so ?

rather it is printing 'custom test' which is the doc of the module node.

tanishq67 replied:
Why my test are working and are legit -
They are passing `classdef` Node in the `visit_classdef()` proof - 
![Screenshot from 2021-04-15 12-12-41](https://user-images.githubusercontent.com/52757132/114825653-2263c400-9de4-11eb-89a8-2665b274bf19.png)

tanishq67 replied:
>Regarding what you are expected to do - you should probably first decide the logic that should be used inside visit_classdef function so that both the old checks work as well as the new check can be added, and then modify old tests/write new tests according to that logic. Is it clear now?

The current method in `visit_classdef()` is not so simple to understand **but works perfectly**, that's why I am trying to fix its unit tests.

tanishq67 replied:
@sajalasati Please let me know your thoughts. Thanks!

tanishq67 replied:
>And how it it possible that those tests were passing earlier if they were passing the wrong node only? Go through the old code again and make sure you properly understand how it works.

![Screenshot from 2021-04-15 12-42-52](https://user-images.githubusercontent.com/52757132/114828980-2691e080-9de8-11eb-9032-0698a25ae8ce.png)

As the previous code was processing the raw file ( it is using `node.root().file` ) and searching for `"'`, it will reach the class's docstring and hence, unit tests will pass.
Also, an edge case which is not being taken of, is if in a test I wrote two classes in the same file, it will only check the first one. ( this will be covered if we will pass `classdef node` to `visit_classdef()`)

sajalasati replied:
@tanishq67 Thanks for explaning the difference in outputs, I get it now. Let me know once you have fixed the checker code (based on my suggestions [here](https://github.com/oppia/oppia/pull/12279#discussion_r606631926)), I will take another full pass on this PR. Thanks!

tanishq67 replied:
@sajalasati There's one more edge case not handled by the author of the PR- 

![Screenshot from 2021-04-17 09-28-54](https://user-images.githubusercontent.com/52757132/115101157-7646e800-9f5f-11eb-9162-82fea225aeb4.png)

How, can we assume that it will have only one decorator ?
Moreover, any decorator can exceed 80 word limit as well, hence it will take up more than a line.

I think the current solution is handling all the edge cases (may be that's why the PR previously got merged) , Please let me know your thoughts so that I can complete the issue.

sajalasati replied:
@tanishq67 I didn't mean that the author's solution is perfect, but you can draw inspiration from it as to how to avoid reading lines from file. Just for providing better context, can you point us to examples of these in our codebase if you found any. (It's okay if you didn't find any, I am just asking)

> How, can we assume that it will have only one decorator ?

@DubeySandeep Do you think this is something @tanishq67 should consider? After looking at some files (in `controllers/`), I am not sure if we promote use of multiple decorators.

> Moreover, any decorator can exceed 80 word limit as well, hence it will take up more than a line.

@tanishq67 Can we use a logic similar to `node.args.args` or `node.doc` (whichever works, or thinking around similar lines) for incrementing the line number accordingly?

tanishq67 replied:
>@tanishq67 I didn't mean that the author's solution is perfect, but you can draw inspiration from it as to how to avoid reading lines from file. Just for providing better context, can you point us to examples of these in our codebase if you found any. (It's okay if you didn't find any, I am just asking)

@sajalasati I looked at couple of different files but didn't found any case.

tanishq67 replied:
>@DubeySandeep Do you think this is something @tanishq67 should consider? After looking at some files (in controllers/), I am not sure if we promote use of multiple decorators.

@DubeySandeep

sajalasati commented on scripts/linters/pylint_extensions.py:
```diff
@@ -590,14 +590,14 @@ class DocstringParameterChecker(checkers.BaseChecker):
     def visit_classdef(self, node):
-        """Visit each class definition in a module and check if there is a
-        single new line below each class docstring.
+        """Visit each class definition and check if there is a
```
The docstring should denote what exactly is going to be done here, like what all checks would be done or whatever description seems appropriate. This is not about just one check, right?

tanishq67 replied:
Oh yes, my bad I forgot to edit it. Thanks!

sajalasati commented on scripts/linters/pylint_extensions.py:
```diff
@@ -619,6 +619,7 @@ def visit_classdef(self, node):
             self.add_message('newline-below-class-docstring', node=node)
         elif second_line_after_doc == b'':
             self.add_message('newline-below-class-docstring', node=node)
+        self.check_docstring_structure(node)
```
Few things:
1) Can we get away with the above 2 checks (if the same are covered in `check_docstring_structure`)?
2) There are other places where we do call this function again for classes having a constructor method - leading to repeated check of the same thing. Please think about how we can avoid that.

tanishq67 replied:
I dont think we can get away with them as their not being covered in the `check_docstring_strucutre` also I dont think adding them to `check_docstring_strucutre` will be a good option either as they are class specific and `check_docstring_strucutre` is used for the fucntiondef node as well.

Are you taking `check_docstring_style` ?? if yes then its deleted in this PR as it is of no use now as its work is directly being done by the passing `classdef node` to `check_docstring_strucutre`.

sajalasati replied:
> I dont think we can get away with them as their not being covered in the `check_docstring_strucutre` also I dont think adding them to `check_docstring_strucutre` will be a good option either as they are class specific and `check_docstring_strucutre` is used for the fucntiondef node as well.

Okay, thanks! I just wanted you to check once. Actually there is PR currently trying to disallow new line after function docstrings so yeah we don't need to do it for functions.

> Are you taking `check_docstring_style` ?? if yes then its deleted in this PR as it is of no use now as its work is directly being done by the passing `classdef node` to `check_docstring_strucutre`.

Where is that deleted part which was doing checks for class docstrings? I don't see the change in this PR yet.

tanishq67 replied:
let me check I thought I have deleted it.

sajalasati commented on scripts/linters/pylint_extensions.py:
```diff
@@ -590,14 +590,14 @@ class DocstringParameterChecker(checkers.BaseChecker):
 
         Args:
-            node: astroid.nodes.ClassDef. Node for a class definition
+            node: astroid.nodes.Classdef. Node for a class definition
```
Please revert this change. The earlier name was correct.

tanishq67 replied:
Yeah sure!

sajalasati commented on scripts/linters/pylint_extensions.py:
```diff
@@ -590,14 +590,14 @@ class DocstringParameterChecker(checkers.BaseChecker):
                 in the AST.
         """
-        # Check if the given node has docstring.
+
```
Remove this extra line.

sajalasati reviewed (changes requested):
Thanks @tanishq67 for explaining everything in detail! Took another pass and left some comments, PTAL!

sajalasati commented:
@tanishq67 Left replies to some of your comments. Don't forget to also address the remaining parts of those comments before next review. Thanks!

sajalasati commented:
Hi @tanishq67! Are you working on this PR currently? If yes then when can you provide an update? We have this issue on our priority list currently, so I can take this up if you want. Kindly let me know. Thanks!

tanishq67 commented:
Sorry, for the delay. I will update today, and soon it will be ready to be merged

DubeySandeep commented:
@tanishq67 Any update on this PR?

tanishq67 commented:
> @tanishq67 Any update on this PR?

@DubeySandeep I got stuck due to #12537. However, now its all working smooth. Please take a look at [this](https://github.com/oppia/oppia/pull/12279#discussion_r615299543)

tanishq67 commented:
> Unassigning @tanishq67 since a re-review was requested. @tanishq67, please make sure you have addressed all review comments. Thanks!

Re assigned  myself  since the PR is not complete yet, just needed  clarification

tanishq67 commented:
> > Unassigning @tanishq67 since a re-review was requested. @tanishq67, please make sure you have addressed all review comments. Thanks!
> 
> Re assigned myself since the PR is not complete yet, just needed a clarification

DubeySandeep commented:
> @DubeySandeep I got stuck due to #12537. However, now its all working smooth. Please take a look at this

@tanishq67 Do you need any help or it's working fine now? Do you want me to review this PR?

tanishq67 commented:
@DubeySandeep PTAL can you please review the PR once and tell me your thoughts on it?
Since I remember the PR was good but we were trying to find a way to remove that `while` loop. 
 I am really puzzled due to the pandemic and on top of that I have my end semester examinations as well.

DubeySandeep commented:
> Since I remember the PR was good but we were trying to find a way to remove that while loop.

@tanishq67  I won't be able to help you with only this much info as I'm not aware of the situation/thread. Can you please give more context, maybe a link to the thread?

tanishq67 commented:
> > Since I remember the PR was good but we were trying to find a way to remove that while loop.
> 
> @tanishq67 I won't be able to help you with only this much info as I'm not aware of the situation/thread. Can you please give more context, maybe a link to the thread?

PTAL @DubeySandeep I am extremely sorry for not giving you context, 
This [comment](https://github.com/oppia/oppia/pull/12279#issuecomment-812754913) covers entire detail about the PR. 

Also, the major issue/discussion was that we were trying to keep the tests testing the lint test simple. 
[Thread](https://github.com/oppia/oppia/pull/12279#discussion_r606631926) 

Let me know what you think - does current tests (unit test testing the lint test) (currently in the PR) does the job for the codebase Or if not then if possible suggest, some alternative since I was stuck while trying to do the same( that thread will give you context)

tanishq67 commented:
@DubeySandeep PTAL

tanishq67 commented:
Please refer [this](https://github.com/oppia/oppia/pull/12279#issuecomment-832306708)

DubeySandeep commented:
@tanishq67, Thanks for your question, I have some quick follow-up questions and I wanted to discuss a few more things related to the issue, I have sent you an invite on hangout can you please accept the invite so that we can quickly resolve this issue?
